templates: migrate Research Group template from v1 to v2 blocks

Agos95 opened the pull request:
### Purpose

Content of this PR:
- update research group template to block v2 system (fix #2908)
- port Slider block to the v2 collection
  - parameters for the background of each slide have been aligned to the ones of [section background](https://wowchemy.com/docs/getting-started/page-builder/#background), with the exception of the video background
- text content of the Contact block is put inside a `<p></p>` tag, to preserve the correct spacing between the text and the contact form
- port People (aka Team Members) block to v2 collection

### Documentation

Need to create two entries in the [blocks](https://wowchemy.com/blocks/) documentation for the v2 version of Slider and People.

Updated versions of the code blocks in the docs are:
- **Slider**
```yaml
---
title: My Page

type: landing

sections:
  - block: slider
    design:
      # Slide height is automatic unless you force a specific height (e.g. '400px')
      slide_height: ''
      is_fullscreen: true
      # Automatically transition through slides?
      loop: false
      # Duration of transition between slides (in ms)
      interval: 2000
    content:
      slides:
      - title: 👋 Welcome to the group
        content: Take a look at what we're working on...
        align: center
        background:
          image:
            filename: coders.jpg
            filters:
              brightness: 0.7
          position: right
          color: '#666'
      - title: Lunch & Learn ☕️
        content: 'Share your knowledge with the group and explore exciting new topics together!'
        align: left
        background:
          image:
            filename: contact.jpg
            filters:
              brightness: 0.7
          position: center
          color: '#555'
      - title: World-Class Semiconductor Lab
        content: 'Just opened last month!'
        align: right
        background:
          image:
            filename: welcome.jpg
            filters:
              brightness: 0.5
          position: center
          color: '#333'
        link:
          icon: graduation-cap
          icon_pack: fas
          text: Join Us
          url: ../contact/
---
```
- **People**
```yaml
---
title: My Page
type: landing

sections:
  - block: people
    content:
      title: Meet the Team
      # Choose which groups/teams of users to display.
      #   Edit `user_groups` in each user's profile to add them to one or more of these groups.
      user_groups:
          - Principal Investigators
          - Researchers
          - Grad Students
          - Administration
          - Visitors
          - Alumni
      # # Field to sort by; one of: last_name (default), first_name, title
      sort_by: last_name
      sort_ascending: true
    design:
      # Show user's social networking links? (true/false)
      show_social: false
      # Show user's interests? (true/false)
      show_interests: true
      # Show user's role?
      show_role: true
      # Show user's organizations/affiliations?
      show_organizations: true
---
```

gcushen commented on starters/research-group/go.mod:
```diff
@@ -5,5 +5,5 @@ go 1.15
 	github.com/wowchemy/wowchemy-hugo-themes/modules/wowchemy-plugin-netlify v1.0.0 // indirect
 	github.com/wowchemy/wowchemy-hugo-themes/modules/wowchemy-plugin-netlify-cms v1.0.0 // indirect
-	github.com/wowchemy/wowchemy-hugo-themes/modules/wowchemy/v5 v5.6.0 // indirect
+	github.com/wowchemy/wowchemy-hugo-themes/modules/wowchemy/v5 main // indirect
```
This should be a specific version (commit or tag), not 'main' which will cause the site to break when contributors update Wowchemy.

Agos95 replied:
Which version should I put? 
Because the latest one is `v5.7.1-0.20230106134435-b9752d697a7b`, which does not include the `people` and `slider` v2 blocks, which are required for the updated version of the Research Group template.

gcushen replied:
Yes, usually we would update the Wowchemy module in a dedicated PR first and then follow up with a PR to update the version of Wowchemy which templates use (as well as applying any breaking changes to the templates in the upgrade process).

In this case since you did a single PR to both add new features and update the templates, we can leave it as main, but please follow up after this PR is merged to fix the version to a specific version, otherwise everyone's site can break when contributors submit PRs with breaking changes to the main branch.

gcushen commented on starters/research-group/content/people/index.md:
```diff
@@ -1,4 +1,21 @@
 ---
```
Missing page title and date

Agos95 commented on modules/wowchemy/layouts/partials/blocks/people.html:
```diff
@@ -0,0 +1,66 @@
+  {{ end }}
+
+  {{ range $block.content.user_groups }}
+  {{ $query := where (where site.Pages "Section" "authors") ".Params.user_groups" "intersect" (slice .) }}
```
Do you prefer if I add a `sort_by` parameter in the people block (along with a `sort_scending`), which defaults to `last_name`, in the same way as in the `collection` block? 
In this way a user can decide whether to sort them by `title`, `first_name` or `last_name`.

gcushen replied:
Good idea, let's try align the People sort options with the similar logic from the Collection block.

Note: I forget the exact logic we have in the Collections block, but that might mean users have to input `Params.last_name` to sort by last name as opposed to just `last_name` since Hugo accesses its own page parameters differently to custom parameters.

Agos95 replied:
I set that users need to input one of `last_name` (default), `first_name` or `title`. In the template then I'll add `.Params.*`.
If something else is passed as argument, it defaults to `last_name`.

Agos95 commented on modules/wowchemy/layouts/partials/blocks/people.html:
```diff
@@ -0,0 +1,66 @@
+  {{ with $block.content.title }}
+  <div class="col-md-12 section-heading">
+    <h1>{{ . | markdownify | emojify }}</h1>
+    {{ if $block.content.subtitle }}<p>{{ $block.content.subtitle | markdownify | emojify }}</p>{{ end }}
```
My bad, I just copied the old v1 widget without noticing these changes.
In the v1 version, the `people` widget did not inherit the heading from the common code in `parse_block_v1`:

https://github.com/wowchemy/wowchemy-hugo-themes/blob/b9752d697a7bb94de3d085db3988606e138f6657/modules/wowchemy/layouts/partials/functions/parse_block_v1.html#L111

I think this is due to the fact that this widget was not intended to be use with a 2 columns layout. I tried to modify the code, and the result for the 2 columns layout is the following:

![people-2-columns](https://user-images.githubusercontent.com/43784886/217489106-b841e55f-2210-4796-9db0-107f15d5e845.png)

which I personally do not really like due to the huge blank space on the left.

I can easily achieve both solutions (either the common block heading with the possibility of 2 columns layout or the specific one inside `people.html`, with only the possibility of 1 column): tell me what you think would be the best one.

gcushen replied:
OK, let's just update the HTML and styles for the single column design in the PR so that they are consistent with the one column wrapper in that function, as per the first comment above.

Agos95 commented:
Added the requested changes. In addition, I also substituted `markdownify` with `$page.RenderString` for `block.content.text`, to be compliant with #2919.

Also, I updated the first comment with the new parameters for the `people` block.

As soon as the PR will be merged, I'll send another one to fix the `go.mod` file in the research group.

gcushen commented:
@Agos95 please resolve the merge conflicts (contact.html)

Agos95 commented:
Solved!

gcushen commented on modules/wowchemy/layouts/partials/blocks/people.html:
```diff
@@ -0,0 +1,74 @@
+
+    {{/* Sort */}}
+    {{ $sort_by := $block.content.sort_by | default "last_name" }}
+    {{ if not ( in "last_name first_name title" $sort_by ) }} {{ $sort_by = "last_name" }} {{ end }}
```
Although limiting the sort options can be useful to prevent sorting on invalid options, we should probably just remove this line and let users sort on anything they want, otherwise, we will probably just get feedback from users trying to search on custom parameters and complaining it doesn't work.

Agos95 replied:
Ok, I understand that and I am going to remove this line

gcushen commented on modules/wowchemy/layouts/partials/blocks/people.html:
```diff
@@ -0,0 +1,74 @@
+    {{/* Sort */}}
+    {{ $sort_by := $block.content.sort_by | default "last_name" }}
+    {{ if not ( in "last_name first_name title" $sort_by ) }} {{ $sort_by = "last_name" }} {{ end }}
+    {{ $sort_by = printf ".Params.%s" $sort_by }}
```
If we prefix `.Params.` to the `sort_by` automatically, we should probably do the same for any other uses of the sort_by option in Wowchemy, such as in the Collections (and Portfolio?) blocks, otherwise different behaviours will be very confusing to users.

Agos95 replied:
Ok, I also remove the automatic `.Params.` prefix to align this block to the others.
Although, I think maybe it could be nice to have an autoprefix just if `last_name`, `first_name` or `title` are passed as parameters, since these are the most common options to sort the people.

It could be something like:
```
{{ if in "last_name first_name title" $sort_by }} {{ $sort_by = printf ".Params.%s" $sort_by }}  {{ end }}
```

What do you think? Otherwise, I'll just remove it, no problem!

gcushen replied:
If the first letter of the sort_by param is capitalised then it's a built-in Hugo param right, otherwise it's a parameter which requires prefixing with `Params.`?

I have also observed that users find this Hugo behaviour very confusing and prone to error.

So I suggest we check if first letter of sort_by is capitalised, if it is then pass it as it is, otherwise prefix it. If this works successfully, then let's apply this logic to all instances where sort_by is used.

Agos95 replied:
I need to think about this. In principle, I think nothing stops me to create a custom parameter (eg. `LastName` instead of `last_name`), so I don't know if this logic can actually be applied.

If you agree, we can close this PR (which I'll immediately follow with a new one to fix the Wowchemy version in `go.mod`) as it is, so that users need to input `.Params.last_name` to sort authors by their last name. 

Then, I'll open another issue to discuss and investigate about your proposed solution to discover built-in or custom parameters.

gcushen replied:
@Agos95 the Wowchemy parameters are all standardised to lowercase-underscore convention, so that is fine, we can go ahead and make things easier for the user based on the above comment at https://github.com/wowchemy/wowchemy-hugo-themes/pull/2914#discussion_r1111261900

Agos95 replied:
Ok, I added the function in the `wowchemy-core` module, since it is independent on bootstrap, and I tested it locally.
I also add it in `people`, `collection` and `portfolio` blocks as comments. This is due to the fact that the version of `wowchemy-core` must be updated (in `modules/wowchemy/go.mod`) in order to detect the new function, and I don't have any idea on what I should do.

Maybe the best thing is to merge this PR, create a new tag/version for `wowchemy-core`, and then I can follow with another PR to fix the versions. But I wait for your instructions, since I don't have any experience on this.

gcushen replied:
See https://github.com/wowchemy/wowchemy-hugo-themes/pull/2914#issuecomment-1505706541

In order to merge this PR without splitting it into separate PRs for each module change, we will go ahead with the latter suggestion you made.

gcushen commented:
@Agos95 You have attempted to modify 3 Go Modules which depend on each other in a single PR.

However, Hugo needs us specify in each module's go.mod which versions of modules it depends on and we cannot retrieve the version (Git commit ref) until we have merged the PR.

So in future, we should generally submit a separate PR for each module when improvements are made to multiple inherited modules.

To help facilitate merging this, please rebase the branch on the latest "main", uncomment the commented code (it's bad practice for us to commit commented code), and set the dependency version of each module to main.

Then once we merge, please submit an update as soon as you can to fix each module's go.mod to require a specific version.

Also, feel free to feedback to Hugo team if you have suggestions on how they can improve the module system to make it easier for contributors.
